File: gtx-out/src/CardDemo.OnlineTransactionManagement.Application/AddTransactionInput.cs
namespace CardDemo.OnlineTransactionManagement.Application;

/// <summary>Input record for adding a new transaction.</summary>
public sealed record AddTransactionInput
{
    /// <summary>Account identifier. Either this or CardNumber must be provided.</summary>
    public long? AccountId { get; init; }

    /// <summary>16-character card number. Either this or AccountId must be provided.</summary>
    public string? CardNumber { get; init; }

    /// <summary>Transaction type code (e.g. "01" credit, "02" payment, "03" purchase).</summary>
    public required string TypeCode { get; init; }

    /// <summary>Transaction category code.</summary>
    public required int CategoryCode { get; init; }

    /// <summary>Transaction source.</summary>
    public required string Source { get; init; }

    /// <summary>Transaction description.</summary>
    public required string Description { get; init; }

    /// <summary>Transaction amount as a string; validated for numeric with up to 2 decimal places.</summary>
    public required string Amount { get; init; }

    /// <summary>Date the transaction originated.</summary>
    public required DateTime OriginDate { get; init; }

    /// <summary>Date the transaction is processed.</summary>
    public required DateTime ProcessDate { get; init; }

    /// <summary>Merchant identifier.</summary>
    public string? MerchantId { get; init; }

    /// <summary>Merchant name.</summary>
    public string? MerchantName { get; init; }

    /// <summary>Merchant city.</summary>
    public string? MerchantCity { get; init; }

    /// <summary>Merchant zip code.</summary>
    public string? MerchantZip { get; init; }
}

File: gtx-out/src/main/java/com/grapeup/carddemo/onlinetransactionmanagement/application/AddTransactionInput.java
package com.grapeup.carddemo.onlinetransactionmanagement.application;

import java.util.Objects;

/**
 * Input for the add-transaction use case.
 * <p>
 * The caller supplies either an 11-character account ID or a 16-character card number.
 * Amount is passed as a raw string so the service can validate its format.
 * All merchant fields are optional (some transactions have no physical merchant).
 * </p>
 */
public record AddTransactionInput(
        String accountIdOrCardNumber,
        String typeCode,
        short categoryCode,
        String source,
        String description,
        String amount,
        String originDate,
        String processDate,
        Integer merchantId,
        String merchantName,
        String merchantCity,
        String merchantZip,
        String cardNumber
) {
    public AddTransactionInput {
        Objects.requireNonNull(accountIdOrCardNumber, "accountIdOrCardNumber is required");
        Objects.requireNonNull(typeCode, "typeCode is required");
        Objects.requireNonNull(source, "source is required");
        Objects.requireNonNull(description, "description is required");
        Objects.requireNonNull(amount, "amount is required");
    }
}
